File: multipicker/src/main/java/com/edwardstock/multipicker/picker/ui/FilesFragment.java
package com.edwardstock.multipicker.picker.ui;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.support.constraint.ConstraintSet;
import android.support.transition.AutoTransition;
import android.support.transition.Transition;
import android.support.transition.TransitionManager;
import android.support.transition.TransitionSet;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Surface;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.edwardstock.multipicker.PickerConfig;
import com.edwardstock.multipicker.R;
import com.edwardstock.multipicker.R2;
import com.edwardstock.multipicker.data.Dir;
import com.edwardstock.multipicker.data.MediaFile;
import com.edwardstock.multipicker.internal.MediaFileLoader;
import com.edwardstock.multipicker.internal.PickerSavePath;
import com.edwardstock.multipicker.internal.helpers.DisplayHelper;
import com.edwardstock.multipicker.picker.PickerConst;
import com.edwardstock.multipicker.picker.adapters.FilesAdapter;
import com.edwardstock.multipicker.picker.adapters.MediaFileDetailsLookup;
import com.edwardstock.multipicker.picker.adapters.MediaFileKeyProvider;
import com.edwardstock.multipicker.picker.views.FilesPresenter;
import com.edwardstock.multipicker.picker.views.FilesView;
import com.google.common.collect.ImmutableList;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

import androidx.recyclerview.selection.ItemKeyProvider;
import androidx.recyclerview.selection.SelectionTracker;
import androidx.recyclerview.selection.StorageStrategy;
import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

/**
 * android-multipicker. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */
public class FilesFragment extends PickerFileSystemFragment implements FilesView {

    private final static String LIST_STATE = "LIST_STATE";
    FilesPresenter presenter;
    @BindView(R2.id.list) RecyclerView list;
    @BindView(R2.id.mp_selection_title) TextView selectionTitle;
    @BindView(R2.id.mp_selection_action_clear) ImageView selectionClearAction;
    @BindView(R2.id.mp_selection_action_done) ImageView selectionDoneAction;
    @BindView(R2.id.progress) ProgressBar progress;
    @BindView(R2.id.selection_root_sub) ConstraintLayout rootSub;
    private SelectionTracker<MediaFile> mSelectionTracker;
    private boolean mLastSelectionState = false;
    private Dir mDir;
    private PickerConfig mConfig;
    private List<MediaFile> mSelected = new ArrayList<>(10);
    private MediaFile mLastPreview = null;
    private Parcelable mListState;

    public static FilesFragment newInstance(PickerConfig config, Dir dir) {
        Bundle args = new Bundle();
        args.putParcelable(PickerConst.EXTRA_DIR, dir);
        args.putParcelable(PickerConst.EXTRA_CONFIG, config);

        FilesFragment fragment = new FilesFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        setRetainInstance(true);
        if (presenter != null) {
            presenter.onRestoreSavedState(savedInstanceState);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        presenter = new FilesPresenter();
        presenter.attachToLifecycle(this);
        presenter.attachView(this);
    }


    @Override
    public void setAdapter(final RecyclerView.Adapter<?> adapter) {
        Timber.d("Set adapter from fragment");
        boolean isTablet = getResources().getBoolean(R.bool.mp_isTablet);
        int spanCount = isTablet ? mConfig.getFileColumnsTablet() : mConfig.getFileColumns();
        if (getActivity() != null) {
            int rot = getActivity().getWindowManager().getDefaultDisplay().getRotation();
            if (rot == Surface.ROTATION_90 || rot == Surface.ROTATION_180) {
                spanCount = (isTablet ? mConfig.getFileColumnsTablet() : mConfig.getFileColumns()) + 1;
            } else {
                spanCount = isTablet ? mConfig.getFileColumnsTablet() : mConfig.getFileColumns();
            }
        }

        GridLayoutManager layoutManager = new GridLayoutManager(getContext(), spanCount);
        setGridSpacingItemDecoration(list, spanCount);
        list.setLayoutManager(layoutManager);
        list.setAdapter(adapter);
        list.setItemAnimator(null);
        layoutManager.onRestoreInstanceState(mListState);
        list.addOnScrollListener(new RecyclerView.OnScrollListener() {


            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    ((FilesAdapter) adapter).setIsScrolling(false);
                } else {
                    ((FilesAdapter) adapter).setIsScrolling(true);
                }
            }
        });

        FilesAdapter filesAdapter = ((FilesAdapter) adapter);

        if (mSelectionTracker != null) {
            Timber.d("Reusing tracker");
            filesAdapter.setSelectionTracker(mSelectionTracker);
            return;
        }
        Timber.d("Create tracker");
        mSelectionTracker = new SelectionTracker.Builder<>(
                "picker",
                list,
                new MediaFileKeyProvider(() -> {
                    if (list.getAdapter() == null) {
                        return Collections.emptyList();
                    }
                    return ((FilesAdapter) list.getAdapter()).getItems();
                }, ItemKeyProvider.SCOPE_MAPPED),
                new MediaFileDetailsLookup(list),
                StorageStrategy.createParcelableStorage(MediaFile.class)
        )
                .withSelectionPredicate(new SelectionTracker.SelectionPredicate<MediaFile>() {
                    @Override
                    public boolean canSetStateForKey(@NonNull MediaFile mediaFile, boolean nextState) {
                        return mConfig.getLimit() <= 0 || (mSelected.size() != mConfig.getLimit() || mSelected.contains(mediaFile));
                    }

                    @Override
                    public boolean canSetStateAtPosition(int position, boolean nextState) {
                        return true;
                    }

                    @Override
                    public boolean canSelectMultiple() {
                        return true;
                    }
                })
                .build();

        mSelectionTracker.addObserver(new SelectionTracker.SelectionObserver<MediaFile>() {
            @Override
            public void onItemStateChanged(@NonNull MediaFile key, boolean selected) {
                super.onItemStateChanged(key, selected);
                presenter.setSelection(ImmutableList.copyOf(mSelectionTracker.getSelection().iterator()));
                mSelected.clear();
                mSelected.addAll(ImmutableList.copyOf(mSelectionTracker.getSelection().iterator()));
                setSubtitle();
                final boolean hasSelection = mSelectionTracker.hasSelection();
                if (mLastSelectionState != hasSelection) {
                    mLastSelectionState = hasSelection;
                    filesAdapter.notifyItemRangeChanged(0, adapter.getItemCount());
                }
            }

            @Override
            public void onSelectionRefresh() {
                super.onSelectionRefresh();
            }

            @Override
            public void onSelectionChanged() {
                super.onSelectionChanged();
            }

            @Override
            public void onSelectionRestored() {
                super.onSelectionRestored();
                final boolean hasSelection = mSelectionTracker.hasSelection();
                presenter.setSelection(ImmutableList.copyOf(mSelectionTracker.getSelection().iterator()));

                if (mLastSelectionState != hasSelection) {
                    mLastSelectionState = hasSelection;
                    filesAdapter.notifyItemRangeChanged(0, adapter.getItemCount());

                }

            }
        });

        filesAdapter.setSelectionTracker(mSelectionTracker);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Timber.d("Delete selection tracker");
        mSelectionTracker = null;
    }

    @Override
    public void setSelectionTitle(CharSequence title) {
        selectionTitle.setText(title);
    }

    @Override
    public void setOnSelectionClearListener(View.OnClickListener listener) {
        selectionClearAction.setOnClickListener(listener);
    }

    @Override
    public void setOnSelectionDoneListener(View.OnClickListener listener) {
        selectionDoneAction.setOnClickListener(listener);
    }

    @Override
    public void setSelectionObserver(SelectionTracker.SelectionObserver<MediaFile> observer) {
        mSelectionTracker.addObserver(observer);
    }

    @Override
    public void showEmpty() {

    }

    @Override
    public void startPreview(MediaFile file, View sharedView) {
        mLastPreview = file;
        safeActivity(act -> {
            act.startPreview(mDir, file, sharedView);
        });
    }

    @Override
    public void clearSelection() {
        if (mSelectionTracker != null) {
            mSelectionTracker.clearSelection();
            mLastSelectionState = false;
        }
    }

    @Override
    public void finishWithResult() {
        safeActivity(act -> act.submitResult(ImmutableList.copyOf(mSelectionTracker.getSelection())));
    }

    @Override
    public void showProgress() {
        progress.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgress() {
        progress.setVisibility(View.GONE);
    }

    @Override
    public void setSelectionSubmitEnabled(boolean enabling) {
        safeActivity(act -> {
            act.runOnUiThread(() -> {
                Timber.d("Set selection submit: %b", enabling);
                ConstraintSet set = new ConstraintSet();
                set.clone(rootSub);
                set.setAlpha(R.id.mp_selection_action_clear, enabling ? 1f : 0f);
                set.setAlpha(R.id.mp_selection_action_done, enabling ? 1f : 0f);
                set.setMargin(R.id.mp_selection_title, ConstraintSet.START, enabling ? DisplayHelper.dpToPx(getContext(), 48) : DisplayHelper.dpToPx(getContext(), 16));
                TransitionSet transitionSet = new TransitionSet();
                Transition shortTransition = new AutoTransition();
                shortTransition.setInterpolator(new DecelerateInterpolator());
//                shortTransition.setDuration(300);
                Transition longTransition = new AutoTransition();
                longTransition.setInterpolator(new AccelerateDecelerateInterpolator());
//                longTransition.setDuration(600);
                transitionSet.addTransition(shortTransition);
                transitionSet.addTransition(longTransition);
                transitionSet.setOrdering(TransitionSet.ORDERING_TOGETHER);

                if (enabling) {
                    longTransition.addTarget(R.id.mp_selection_action_clear);
                    longTransition.addTarget(R.id.mp_selection_action_done);
                    shortTransition.addTarget(R.id.mp_selection_title);
                } else {
                    longTransition.addTarget(R.id.mp_selection_title);
                    shortTransition.addTarget(R.id.mp_selection_action_clear);
                    shortTransition.addTarget(R.id.mp_selection_action_done);
                }

                TransitionManager.beginDelayedTransition(rootSub, transitionSet);
                set.applyTo(rootSub);
            });
        });
    }

    @Override
    public void scrollTo(int position) {
//        if(list != null) {
//            list.post(()->{
//                list.smoothScrollToPosition(position);
//            });
//        }
    }

    public void addSelection(MediaFile selection) {
        if (selection != null && mSelectionTracker != null) {
            mSelectionTracker.select(selection);
        }
    }

    @Override
    public void onError(CharSequence error) {
        safeActivity(act -> act.onError(error));
    }

    @Override
    public void onError(Throwable t) {
        safeActivity(act -> act.onError(t));
    }

    @Override
    public void startUpdateFiles() {
        presenter.updateFiles(new MediaFileLoader(getContext()));
    }

    @Override
    public void onStop() {
        super.onStop();
        if (list != null && list.getLayoutManager() != null) {
            mListState = list.getLayoutManager().onSaveInstanceState();
        }

    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {

        if (mSelectionTracker != null) {
            mSelectionTracker.onSaveInstanceState(outState);
        }
        if (presenter != null) {
            presenter.onSaveInstanceState(outState);
        }

        if (list.getLayoutManager() != null) {
            mListState = list.getLayoutManager().onSaveInstanceState();
            outState.putParcelable(LIST_STATE, mListState);
        }

        super.onSaveInstanceState(outState);
    }

    @Override
    public void onPause() {
        if (mConfig.getLimit() > 0) {
            safeActivity(act -> {
                act.toolbar.setSubtitle(null);
            });
        }
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        list.getLayoutManager().onRestoreInstanceState(mListState);
        if (mConfig.getLimit() > 0) {
            setSubtitle();
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.mp_fragment_filesystem, container, false);
        postponeEnterTransition();
        ButterKnife.bind(this, view);
        mDir = getArguments().getParcelable(PickerConst.EXTRA_DIR);
        mConfig = getArguments().getParcelable(PickerConst.EXTRA_CONFIG);
        presenter.handleExtras(getArguments());
        presenter.updateFiles(new MediaFileLoader(getContext()));

        presenter.setOnFileMeasuredListener(mLastPreview, mf -> startPostponedEnterTransition());

        if (mSelectionTracker != null) {
            mSelectionTracker.onRestoreInstanceState(savedInstanceState);
        }


        return view;
    }

    @Override
    public FilesPresenter getPresenter() {
        return presenter;
    }

    @Override
    protected PickerSavePath getCapturedSavePath() {
        if (mDir != null) {
            return new PickerSavePath(mDir.getName());
        }
        return super.getCapturedSavePath();
    }

    private void setSubtitle() {
        if (mConfig.getLimit() > 0) {
            safeActivity(act -> {
                act.toolbar.setSubtitle(String.format(Locale.getDefault(), "%d / %d", mSelectionTracker.getSelection().size(), mConfig.getLimit()));
            });
        }
    }


}

File: multipicker/src/main/java/com/edwardstock/multipicker/picker/ui/FilesFragment.kt
package com.edwardstock.multipicker.picker.ui

import android.annotation.SuppressLint
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.Surface
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.DecelerateInterpolator
import androidx.annotation.PluralsRes
import androidx.constraintlayout.widget.ConstraintSet
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.selection.ItemKeyProvider
import androidx.recyclerview.selection.SelectionTracker
import androidx.recyclerview.selection.StorageStrategy
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.AutoTransition
import androidx.transition.Transition
import androidx.transition.TransitionManager
import androidx.transition.TransitionSet
import com.edwardstock.multipicker.PickerConfig
import com.edwardstock.multipicker.R
import com.edwardstock.multipicker.data.Dir
import com.edwardstock.multipicker.data.MediaFile
import com.edwardstock.multipicker.databinding.MpFragmentFilesystemBinding
import com.edwardstock.multipicker.internal.MediaFileLoader
import com.edwardstock.multipicker.internal.PickerSavePath
import com.edwardstock.multipicker.internal.helpers.DisplayHelper
import com.edwardstock.multipicker.picker.PickerConst
import com.edwardstock.multipicker.picker.adapters.FilesAdapter
import com.edwardstock.multipicker.picker.adapters.MediaFileDetailsLookup
import com.edwardstock.multipicker.picker.adapters.MediaFileKeyProvider
import com.edwardstock.multipicker.picker.getParcelableCompat
import com.edwardstock.multipicker.picker.views.BaseFsViewModel
import com.edwardstock.multipicker.picker.views.FilesView
import com.edwardstock.multipicker.picker.views.FilesViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import timber.log.Timber
import java.io.File
import java.util.*

/**
 * android-multipicker. 2018
 * @author Eduard Maximovich [edward.vstock@gmail.com]
 */
class FilesFragment : PickerFileSystemFragment(), FilesView {
    private val viewModel: FilesViewModel by viewModels()

    private lateinit var b: MpFragmentFilesystemBinding
    private var selectionTracker: SelectionTracker<MediaFile>? = null
    private var lastSelectionState = false
    private var choosenDir: Dir? = null
    private lateinit var config: PickerConfig
    private val selectedFiles: MutableList<MediaFile> = ArrayList<MediaFile>(10)
    private var lastPreview: MediaFile? = null
    private var listState: Parcelable? = null


    private lateinit var adapter: FilesAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel.onRestoreSavedState(savedInstanceState)
    }

    override fun onStop() {
        super.onStop()
        if (b.list.layoutManager != null) {
            listState = b.list.layoutManager!!.onSaveInstanceState()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        if (selectionTracker != null) {
            selectionTracker!!.onSaveInstanceState(outState)
        }
        viewModel.onSaveInstanceState(outState)
        if (b.list.layoutManager != null) {
            listState = b.list.layoutManager!!.onSaveInstanceState()
            outState.putParcelable(LIST_STATE, listState)
        }
        super.onSaveInstanceState(outState)
    }

    override fun onPause() {
        if (config.limit > 0) {
            safeActivity { act: PickerActivity -> act.b.toolbar.subtitle = null }
        }
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        b.list.layoutManager!!.onRestoreInstanceState(listState)
        if (config.limit > 0) {
            setSubtitle()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        b = MpFragmentFilesystemBinding.inflate(inflater, container, false)

        choosenDir = requireArguments().getParcelableCompat(PickerConst.EXTRA_DIR)
        config = requireArguments().getParcelableCompat(PickerConst.EXTRA_CONFIG)!!

        viewModel.handleExtras(requireArguments())
        if (selectionTracker != null) {
            selectionTracker!!.onRestoreInstanceState(savedInstanceState)
        }
        return b.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        adapter = FilesAdapter(config, null) { file: MediaFile, _: Boolean, sharedView: View ->
//            onFileClick(file, isSelected, sharedView)
            startPreview(file, sharedView)
        }
        initAdapter(adapter)


        viewModel.filesToDelete
                .onEach {
                    it.forEach { file -> removeFileFromMediaDB(file) }
                }
                .launchIn(lifecycleScope)

        viewModel.files
                .onEach {
                    adapter.setData(it)
                    adapter.notifyItemRangeChanged(0, it.size)
                }
                .launchWhileVisible()

        viewModel.showProgress
                .onEach { if (it) showProgress() else hideProgress() }
                .launchWhileVisible()

//        viewModel.showEmpty
//                .onEach { if(it) showEmpty() else hideEmpty() }
//                .launchWhileVisible()

        viewModel.showError
                .onEach {
                    Timber.e(it)
                }
                .launchWhileVisible()

        viewModel.scrollTo
                .onEach {
                    b.list.layoutManager?.scrollToPosition(it)
                }
                .launchWhileVisible()

        viewModel.selectionSubmitEnabled
                .onEach { setSelectionSubmitEnabled(it) }
                .launchWhileVisible()

        viewModel.selectionTitle
                .onEach {
                    when {
                        it.plural == -1 -> setSelectionTitle(it.title)
                        it.plural > -1 -> setSelectionTitleN(it.title, it.plural)
                        it.title2 != -1 && it.plural2 != -1 -> this.setSelectionTitlePhotosAndVideos(it.title, it.plural, it.title2, it.plural2)
                    }
                }
                .launchWhileVisible()

        viewModel.updateFiles(MediaFileLoader(requireContext()))

        setOnSelectionClearListener { clearSelection() }
        setOnSelectionDoneListener { finishWithResult() }
    }

    override fun getViewModel(): BaseFsViewModel {
        return viewModel
    }

    fun initAdapter(adapter: FilesAdapter) {
        Timber.d("Set adapter from fragment")
        val isTablet = resources.getBoolean(R.bool.mp_isTablet)
        var spanCount: Int = if (isTablet) config.fileColumnsTablet else config.fileColumns
        if (activity != null) {
            val rot = DisplayHelper.getRotation(requireActivity())
            spanCount = if (rot == Surface.ROTATION_90 || rot == Surface.ROTATION_180) {
                (if (isTablet) config.fileColumnsTablet else config.fileColumns) + 1
            } else {
                if (isTablet) config.fileColumnsTablet else config.fileColumns
            }
        }
        val layoutManager = GridLayoutManager(context, spanCount)
        setGridSpacingItemDecoration(b.list, spanCount)
        b.list.layoutManager = layoutManager
        b.list.adapter = adapter
        b.list.itemAnimator = null
        layoutManager.onRestoreInstanceState(listState)
        b.list.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                adapter.setIsScrolling(newState != RecyclerView.SCROLL_STATE_IDLE)
            }
        })
        val filesAdapter: FilesAdapter = adapter
        if (selectionTracker != null) {
            Timber.d("Reusing tracker")
            filesAdapter.setSelectionTracker(selectionTracker)
            return
        }
        Timber.d("Create tracker")
        selectionTracker = SelectionTracker.Builder(
                "picker",
                b.list,
                MediaFileKeyProvider({
                    if (b.list.adapter == null) {
                        emptyList()
                    } else {
                        (b.list.adapter as FilesAdapter).items
                    }
                }, ItemKeyProvider.SCOPE_MAPPED),
                MediaFileDetailsLookup(b.list),
                StorageStrategy.createParcelableStorage(MediaFile::class.java)
        )
                .withSelectionPredicate(object : SelectionTracker.SelectionPredicate<MediaFile>() {
                    override fun canSetStateForKey(mediaFile: MediaFile, nextState: Boolean): Boolean {
                        return config.limit <= 0 || selectedFiles.size != config.limit || selectedFiles.contains(mediaFile)
                    }

                    override fun canSetStateAtPosition(position: Int, nextState: Boolean): Boolean {
                        return true
                    }

                    override fun canSelectMultiple(): Boolean {
                        return true
                    }
                })
                .build()

        selectionTracker?.addObserver(object : SelectionTracker.SelectionObserver<MediaFile?>() {
            override fun onItemStateChanged(key: MediaFile, selected: Boolean) {
                super.onItemStateChanged(key, selected)
                viewModel.setSelection(selectionTracker!!.selection.toList())
                selectedFiles.clear()
                selectedFiles.addAll(selectionTracker!!.selection.toList())
                setSubtitle()
                val hasSelection: Boolean = selectionTracker!!.hasSelection()
                if (lastSelectionState != hasSelection) {
                    lastSelectionState = hasSelection
                    filesAdapter.notifyItemRangeChanged(0, adapter.itemCount)
                }
            }

            override fun onSelectionRestored() {
                super.onSelectionRestored()
                val hasSelection: Boolean = selectionTracker!!.hasSelection()
                viewModel.setSelection(selectionTracker!!.selection.toList())
                if (lastSelectionState != hasSelection) {
                    lastSelectionState = hasSelection
                    filesAdapter.notifyItemRangeChanged(0, adapter.itemCount)
                }
            }
        })
        filesAdapter.setSelectionTracker(selectionTracker)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        Timber.d("Delete selection tracker")
        selectionTracker = null
    }

    override fun setSelectionTitle(title: CharSequence) {
        b.mpSelectionTitle.text = title
    }

    override fun setSelectionTitle(titleRes: Int) {
        b.mpSelectionTitle.setText(titleRes)
    }

    override fun setSelectionTitleN(@PluralsRes titleRes: Int, n: Int) {
        b.mpSelectionTitle.text = requireContext().resources.getQuantityString(titleRes, n, n)
    }

    @SuppressLint("SetTextI18n")
    override fun setSelectionTitlePhotosAndVideos(@PluralsRes prefix: Int, np: Int, @PluralsRes suffix: Int, ns: Int) {
        val res = requireContext().resources
        b.mpSelectionTitle.text = "${res.getQuantityString(prefix, np, np)}${res.getQuantityString(suffix, ns, ns)}"
    }

    override fun setOnSelectionClearListener(listener: View.OnClickListener) {
        b.mpSelectionActionClear.setOnClickListener(listener)
    }

    override fun setOnSelectionDoneListener(listener: View.OnClickListener) {
        b.mpSelectionActionDone.setOnClickListener(listener)
    }

    override fun setSelectionObserver(observer: SelectionTracker.SelectionObserver<MediaFile>) {
        selectionTracker!!.addObserver(observer)
    }

    override fun startPreview(file: MediaFile, sharedView: View) {
        lastPreview = file
        safeActivity { act: PickerActivity -> act.startPreview(choosenDir, file, sharedView) }
    }

    override fun clearSelection() {
        if (selectionTracker != null) {
            selectionTracker!!.clearSelection()
            lastSelectionState = false
        }
    }

    override fun finishWithResult() {
        safeActivity { act: PickerActivity ->
            act.submitResult(selectionTracker?.selection?.toList() ?: emptyList())
        }
    }

    override fun showProgress() {
        b.progress.visibility = View.VISIBLE
    }

    override fun hideProgress() {
        b.progress.visibility = View.GONE
    }

    override fun setSelectionSubmitEnabled(enabled: Boolean) {
        safeActivity { act: PickerActivity ->
            act.runOnUiThread {
                Timber.d("Set selection submit: %b", enabled)
                val set = ConstraintSet()
                set.clone(b.selectionRootSub)
                set.setAlpha(R.id.mp_selection_action_clear, if (enabled) 1f else 0f)
                set.setAlpha(R.id.mp_selection_action_done, if (enabled) 1f else 0f)
                set.setMargin(R.id.mp_selection_title, ConstraintSet.START, if (enabled) DisplayHelper.dpToPx(requireContext(), 48) else DisplayHelper.dpToPx(requireContext(), 16))
                val transitionSet = TransitionSet()
                val shortTransition: Transition = AutoTransition()
                shortTransition.interpolator = DecelerateInterpolator()
                //                shortTransition.setDuration(300);
                val longTransition: Transition = AutoTransition()
                longTransition.interpolator = AccelerateDecelerateInterpolator()
                //                longTransition.setDuration(600);
                transitionSet.addTransition(shortTransition)
                transitionSet.addTransition(longTransition)
                transitionSet.ordering = TransitionSet.ORDERING_TOGETHER
                if (enabled) {
                    longTransition.addTarget(R.id.mp_selection_action_clear)
                    longTransition.addTarget(R.id.mp_selection_action_done)
                    shortTransition.addTarget(R.id.mp_selection_title)
                } else {
                    longTransition.addTarget(R.id.mp_selection_title)
                    shortTransition.addTarget(R.id.mp_selection_action_clear)
                    shortTransition.addTarget(R.id.mp_selection_action_done)
                }
                TransitionManager.beginDelayedTransition(b.selectionRootSub, transitionSet)
                set.applyTo(b.selectionRootSub)
            }
        }
    }

    override fun removeFileFromMediaDB(file: File) {
        safeActivity { act: PickerActivity ->
            Timber.d("Remove file: %s as it doesn't exists", file.absoluteFile)
            try {
                act.contentResolver.delete(Uri.fromFile(file), null, null)
                file.delete()
            } catch (ignore: Throwable) {
            }
        }
    }

    fun addSelection(selection: MediaFile?) {
        if (selection != null && selectionTracker != null) {
            selectionTracker!!.select(selection)
        }
    }

    override fun onError(error: CharSequence?) {
        safeActivity { act: PickerActivity -> act.onError(error) }
    }

    override fun onError(t: Throwable?) {
        safeActivity { act: PickerActivity -> act.onError(t) }
    }

    override val capturedSavePath: PickerSavePath?
        get() = if (choosenDir != null) {
            PickerSavePath(choosenDir!!.name!!)
        } else {
            super.capturedSavePath
        }

    private fun setSubtitle() {
        if (config.limit > 0) {
            safeActivity { act: PickerActivity -> act.b.toolbar.subtitle = String.format(Locale.getDefault(), "%d / %d", selectionTracker!!.selection.size(), config.limit) }
        }
    }

    companion object {
        private const val LIST_STATE = "LIST_STATE"
        fun newInstance(config: PickerConfig?, dir: Dir?): FilesFragment {
            val args = Bundle()
            args.putParcelable(PickerConst.EXTRA_DIR, dir)
            args.putParcelable(PickerConst.EXTRA_CONFIG, config)
            val fragment = FilesFragment()
            fragment.arguments = args
            return fragment
        }
    }
}